File: lib/services/grouping_engine.dart
import 'dart:core';
import '../models/sms_message.dart';
import '../models/group_rule.dart';

class GroupingEngine {
  static const GroupingEngine instance = GroupingEngine._internal();
  factory GroupingEngine() => instance;
  GroupingEngine._internal();

  String? processMessage(SMSMessage message, List<GroupRule> rules) {
    final enabledRules = rules
        .where((rule) => rule.isEnabled)
        .toList()
      ..sort((a, b) => a.priority.compareTo(b.priority));

    for (final rule in enabledRules) {
      if (_matchesRule(message, rule)) {
        return rule.id;
      }
    }
    return null;
  }

  bool _matchesRule(SMSMessage message, GroupRule rule) {
    switch (rule.matchType) {
      case MatchType.keyword:
        return rule.patterns.any((pattern) =>
          message.content.contains(pattern) ||
          message.sender.contains(pattern)
        );

      case MatchType.regex:
        return rule.patterns.any((pattern) {
          try {
            final regex = RegExp(pattern);
            return regex.hasMatch(message.content) ||
                   regex.hasMatch(message.sender);
          } catch (e) {
            return false;
          }
        });

      case MatchType.sender:
        return rule.patterns.contains(message.sender);
    }
  }

  bool testRule(GroupRule rule, String testText) {
    final testMessage = SMSMessage(
      id: 'test',
      sender: '测试',
      content: testText,
      date: DateTime.now(),
    );
    return _matchesRule(testMessage, rule);
  }

  void processMessages(
    List<SMSMessage> messages,
    List<GroupRule> rules,
    Map<String, List<SMSMessage>> groupedMessages,
  ) {
    groupedMessages.clear();
    
    for (final message in messages) {
      final groupId = processMessage(message, rules);
      if (groupId != null) {
        groupedMessages.putIfAbsent(groupId, () => []).add(message);
      }
    }
  }
}

File: SMSGroupApp/Services/GroupingEngine.swift
import Foundation

class GroupingEngine {
    static let shared = GroupingEngine()
    
    private init() {}
    
    func processMessage(_ message: SMSMessage, with rules: [GroupRule]) -> UUID? {
        let enabledRules = rules.filter { $0.isEnabled }.sorted { $0.priority < $1.priority }
        
        for rule in enabledRules {
            if matchesRule(message, rule: rule) {
                return rule.id
            }
        }
        
        return nil
    }
    
    private func matchesRule(_ message: SMSMessage, rule: GroupRule) -> Bool {
        switch rule.matchType {
        case .keyword:
            return rule.patterns.contains { pattern in
                message.content.contains(pattern) || message.sender.contains(pattern)
            }
            
        case .regex:
            return rule.patterns.contains { pattern in
                guard let regex = try? NSRegularExpression(pattern: pattern) else {
                    return false
                }
                
                let contentRange = NSRange(message.content.startIndex..., in: message.content)
                let senderRange = NSRange(message.sender.startIndex..., in: message.sender)
                
                return regex.firstMatch(in: message.content, range: contentRange) != nil ||
                       regex.firstMatch(in: message.sender, range: senderRange) != nil
            }
            
        case .sender:
            return rule.patterns.contains(message.sender)
        }
    }
    
    func testRule(rule: GroupRule, testText: String) -> Bool {
        let testMessage = SMSMessage(sender: "测试", content: testText, date: Date())
        return matchesRule(testMessage, rule: rule)
    }
}